### DISCUSSION TITLE
Use of @rp2.asm_pio in a module

### DISCUSSION THREAD

**Discussion by sk8board** [micropython/micropython#16000] [Using MicroPython]
I get the following error when using a state machine in a module.

`NameError: name 'rp2' isn't defined`

The following code works independently but does not work as a module.

``` py
from rp2 import PIO, StateMachine, asm_pio
import time
import random

@rp2.asm_pio(in_shiftdir=0, out_shiftdir=1, autopull=True, pull_thresh=1, autopush=True, push_thresh=1)
def bounce():       # move a single bit from TX FIFO to RX FIFO
    wrap_target()

    out(x, 1)       # shift in one bit from OSR to scratch x
    in_(x, 1)       # shift out one bit from scratch x to ISR
    irq(rel(0))     # set state machine 0 interrupt flag for triggering DMA 0 using sm0.irq()
    
    wrap()

def test(x):
    print(sm0.get())

sm0 = rp2.StateMachine(0, bounce)
sm0.irq(handler=test, trigger=1, hard=True)
sm0.active(1)

def print_random_bits():
    for i in range(0,32):
        sm0.put(random.getrandbits(1))
        time.sleep(0.1)
        
print_random_bits()
```

If I save this code to the RP2040 Pico as `Demo_Module.py`, then in `main.py` I will `import Demo_Module`, I receive the error.

`NameError: name 'rp2' isn't defined`

Is there something special to do when using the @ decorator in a module?

**Comment by mendenm** [comment]
you need to do `import rp2` if you want to use the `rp2.asm_pio()` decorator.  As you have the import, you would just use `asm_pio()` as the decorator, since that is the name you imported.

**Reply by sk8board** [reply]
Thank you for your reply.

This code works with `from rp2 import asm_pio` when using `@rp2.asm_pio` when executing the code standalone. Why don't I get this error when using this code standalone (not as a module)?

**Reply by mendenm** [reply]
The rp2 namespace may already be imported into the top-level module.  It won't be automatically there anywhere else.  I've never really tried what automatically appears, since I always explicitly do the imports, just for clarity.  My rp2 isn't at hand right now to test this.
